sway-ir/src/optimize/simplify_cfg.rs

```diff
@@ -27,6 +27,69 @@ pub fn simplify_cfg(context: &mut Context, function: &Function) -> Result<bool,
         }
         break;
     }
+
+    modified |= unlink_empty_blocks(context, function)?;
+
+    Ok(modified)
+}
+
+fn unlink_empty_blocks(context: &mut Context, function: &Function) -> Result<bool, IrError> {
+    let mut modified = false;
+    let candidates: Vec<_> = function
+        .block_iter(context)
+        .skip(1)
+        .filter_map(|block| {
+            match block.get_terminator(context) {
+                // Except for a PHI and a branch, we don't want anything else.
+                // num_instructions doesn't count PHI though.
+                Some(Instruction::Branch(to_block)) if block.num_instructions(context) <= 1 => {
+                    Some((block, *to_block))
+                }
+                _ => None,
+            }
+        })
+        .collect();
+    for (block, to_block) in candidates {
+        let block_phi = block.get_phi(context);
+        let to_block_phi = to_block.get_phi(context);
+        let mut cur_phi_pairs = if let ValueDatum::Instruction(Instruction::Phi(pairs)) =
+            context.values[block_phi.0].value.clone()
+        {
+            pairs
+        } else {
+            vec![]
+        };
+        let preds: Vec<_> = context.blocks[block.0].predecessors(context).collect();
+        if let ValueDatum::Instruction(Instruction::Phi(to_phi_pairs)) =
+            &mut context.values[to_block_phi.0].value
+        {
+            // In `to_block`, we want to re-route all values coming in from `block`
+            // to be coming in from all of `preds`. If there's already a value coming
+            // in from any of `pred`, then there's a conflict. We bail out.
+            if to_phi_pairs.iter().any(|(block, _)| preds.contains(block)) {
+                continue;
+            }
+            let from_block_pair_idx = to_phi_pairs.iter().position(|(b, _)| *b == block);
+            if let Some(idx) = from_block_pair_idx {
+                let (_, v) = to_phi_pairs[idx];
+                to_phi_pairs.swap_remove(idx);
+                if v == block_phi {
+                    // If the value coming to `to_phi` is `block_phi`, we replace it
+                    // with all the incoming values to `block_phi` itself.
+                    to_phi_pairs.append(&mut cur_phi_pairs);
+                } else {
+                    // Otherwise, it gets `v` from every `pred`.
+                    let v_pred_pairs = preds.iter().map(|b| (*b, v));
+                    to_phi_pairs.extend(v_pred_pairs);
+                }
+                modified = true;
+            } // We don't need to bother if there is no value coming in from block.
```

Comment:
I think this is over complicating the process by trying to update the pairs in place.  Most of the time a block will have at most a handful of predecessors so trying to be overly efficient isn't worth it.

Instead a new PHI could be created based on the old ones.  It should be OK to just delete any reference the the empty block PHI from the successor PHIs since we'll inject the predecessor pairs into the PHI anyway, which is all the empty block PHI would've contained.  So it's two steps - remove the empty block PHI value and then merge the predecessor PHIs with the successor PHIs.  They're essentially sets of pairs.

All the PHI manipulation should be managed in `block.rs` too, rather than poking around directly in the optimisation passes, if possible.  How PHIs working internally shouldn't be exposed.  So a `Block::merge_into_phi(other_phi: Value)` method or something.

It might be worth actually making them `BTreeSet`s instead of `Vec`.

Reply:
>merge the predecessor PHIs with the successor PHIs. 

This is not what I'm doing in both the cases. I'm separating the following two scenarios:

```
block0:
v1 = const 1
br block1

block1:
x = phi(block0: v1)
```
and

```
block0:
v1 = phi(...)
br block1

block1:
x = phi(block0: v1, ...)
```

and also note that in the latter case, I actually remove the pair `(block0, v1)` from `block1`'s PHI before doing the merge.

So I can't see how to best do this in a single function call to a new method for `Block`.

Reply:
No, it's two steps.  Remove `v1` altogether from the successor PHIs because we're merging those in in the second step.  Two methods -- `remove_pred_from_phi()` and `merge_phi()`.  But that's only if the value coming from the empty block is its PHI.

That first example is tricky though.  It'll be empty because `const` isn't an instruction.  OK, I didn't think of that.   So you create a new PHI which is all the pred PHIs except with the values replaced with the `const` and then merge that into the succ PHIs.

Hrmm, let's see:
1. Remove `block0` and `v1` from successor PHIs.
2. Let `new_phi` = `v1` if `v1` is a PHI, else `new_phi` is a new PHI with all the pred blocks but `v1` as the values.
3. Merge `new_phi` with successor PHIs.

Does that work?  No editing in place and all of the gnarly stuff in `block.rs`.

Reply:
When I say merge, I mean something like [BTreeSet::append()](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#method.append) which will handle duplicates.  If the check for conflicts remains then it should work.

Or maybe a new set with `BTreeSet::from_iter(a.union(b))`.  Whichever works.

Reply:
Or even... don't merge, just call `Block::add_phi()` to the successors with the old pairs, I guess.  If we've guaranteed no conflicts then just adding the new ones like you've done with `to_phi_pairs.extend(v_pred_pairs)` above shouldn't create dupes.

I'd just mostly like to see the direct pair manipulation kept in `block.rs`.